core/src/main/java/org/apache/iceberg/rest/http/HttpClient.java

```diff
@@ -0,0 +1,235 @@
+import org.apache.iceberg.rest.RestClient;
+import org.apache.iceberg.rest.UncheckedRestException;
+
+/**
+ * An HttpClient that can be used for the RestClient (eventually).
+ * TODO - Extract out top  for all of this so that other implementations are easy to implement (e.g. gRPC).
+ */
+public class HttpClient implements RestClient {
+  private final String baseUrl;
+
+  // Need their own Builders?
+  private final CloseableHttpClient httpClient;
+  private final Consumer<CloseableHttpResponse> defaultErrorHandler;
+  private final ObjectMapper mapper;
+  private Map<String, String> additionalHeaders = Maps.newHashMap();
+  private Consumer<HttpUriRequest> requestInterceptor = (r) -> { };
+
+  private HttpClient(
+      String baseUrl, CloseableHttpClient httpClient, ObjectMapper mapper,
+      Map<String, String> additionalHeaders, Consumer<HttpUriRequest> requestIntercepter,
+      Consumer<CloseableHttpResponse> defaultErrorHandler) {
+    this.baseUrl = baseUrl;
+    this.httpClient = httpClient != null ? httpClient : HttpClients.createDefault();
+    this.mapper = mapper != null ? mapper : new ObjectMapper();
+    this.additionalHeaders = additionalHeaders != null ? additionalHeaders : Maps.newHashMap();
+    // TODO - Best way to handle Preconditions here.
+    this.requestInterceptor = requestIntercepter;
+    this.defaultErrorHandler = defaultErrorHandler;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Method to execute an HTTP request and process the corresponding response.
+   *
+   * @param method       - HTTP method, such as GET, POST, HEAD, etc.
+   * @param path         - URL path to send the request to
+   * @param body         - Contents of the request body.
+   * @param responseType - Class of the Response type. Needs to have serializer registered with ObjectMapper
+   * @param errorHandler - Error handler delegated for HTTP responses which handles server error responses
+   * @param <T>          - Class type of the response for deserialization. Must be registered with the ObjectMapper.
+   * @return The Response enttity, parsed and converted to its type T
+   * @throws UncheckedIOException - Shouldn't throw this as the requestInterceptor should handle expected cases.
+   * @throws UncheckedRestException - Wraps exceptions to avoid having to use checked exceptions everywhere.
+   */
+  @Nullable
+  public <T> T execute(
+      Method method, String path, Object body, Class<T> responseType,
+      Consumer<CloseableHttpResponse> errorHandler) {
+    HttpUriRequestBase request = new HttpUriRequestBase(method.name(), URI.create(String.format("%s/%s", baseUrl,
+        path)));
+    addRequestHeaders(request);
+
+    if (body != null) {
+      try {
+        request.setEntity(new StringEntity(mapper.writeValueAsString(body)));
+      } catch (JsonProcessingException e) {
+        throw new UncheckedRestException(e, "Failed to write request body");
```

Comment:
I think it would be good to have an exception that distinguishes problems before sending the request, like this one.